### ISSUE TITLE
Creating custom loss function using R 

### ISSUE THREAD

**Issue by daniel-althoff** [mlverse/torch#422]
Hi,

I have been trying to follow this tutorial: https://torch.mlverse.org/docs/articles/getting-started/optim.html 
But use a custom loss function instead.
In this loss function I want R flexibility to pass other variables, ID, etc. However, when I create a custom loss function, lets say:

``loss <- mean((as.array(y_pred) - as.array(y))**2) %>% torch_scalar_tensor()``

It ends up as in the same format as using ``loss <- (y_pred - y)$pow(2)$mean()``, but cannot propagate the error.
I get the following error when trying to backward propagate

``loss$backward()`` results in
Error in (function (self, gradient, retain_graph, create_graph)  : 
  element 0 of tensors does not require grad and does not have a grad_fn
Exception raised from run_backward at ..\..\torch\csrc\autograd\autograd.cpp:80 (most recent call first):
00007FFB3DA1A7B200007FFB3DA1A750 c10.dll!c10::Error::Error [<unknown file> @ <unknown line number>]
00007FFADB0D201700007FFADB0D1940 torch_cpu.dll!torch::autograd::Engine::make_anomaly_metadata [<unknown file> @ <unknown line number>]
00007FFADB0D16D600007FFADB0D1660 torch_cpu.dll!torch::autograd::backward [<unknown file> @ <unknown line number>]
00007FFADB4753B400007FFADB4750C0 torch_cpu.dll!torch::autograd::VariableType::allCUDATypes [<unknown file> @ <unknown line number>]
00007FFADB475E5200007FFADB4750C0 torch_cpu.dll!torch::autograd::VariableType::allCUDATypes [<unknown file> @ <unknown line number>]
00007FFAD9F2FB8900007FFAD9F26BA0 torch_cpu.dll!at::is_custom_op [<unknown file> @ <unknown line number>]
00007FFAD9F373F500007FFAD9F26BA0 torch_cpu.dll!at::is_custom_op [<unknown file> @ <unknown line 

Can anyone help me figuring this out how to convert loss calculated in R into an acceptable one?

**Comment by dfalbel** [comment]
It's not possible to use arbitrary R functions. If you want to backpropagate, all operations that you do on tensors must be implemented in torch, without the need for you to convert to R and back to a tensor. 

Depending on what you are doing, you can also write a custom operation, but in this case you also need to implement the how to compute its derivatives, see the extending autograd guide: https://torch.mlverse.org/docs/articles/extending-autograd.html

**Comment by daniel-althoff** [comment]
Thank you @dfalbel 

I am not too familiar with operations with tensors. Could you help me with a custom loss function? 
I work with hydrological modeling, and wanted to train a single model for multiple gauge stations.
Lets say I wanted to decrease the median MSE across multiple gauge stations.
How would I pass the ID gauge stations to the loss function, and how do I group MSE calculations by station to later provide the median?

**Comment by daniel-althoff** [comment]
Thank you @dfalbel.

I looked into the guide, however, I could not figure how to make calculations for grouped observations. 
For example, lets say I have an ID column in my data set which refers to the ID of gauge stations. I wanted to pass the ID for the loss function to be able to compute the MSE for each gauge station and the median among all.
Similar to what is done when 
``dataframe %>% group_by(ID) %>% summarise(MSE = MSE(y_hat, y_true)) %>% ungroup() %>% summarise(MSE = median(MSE))``
  is used in data manipulation.

**Comment by dfalbel** [comment]
Assuming that each group has the same number of observations I would do something like this:

```
y_hat <- torch_randn(100, 10)
y <- torch_randn(100, 10)

loss <- torch_median(torch_mean((y - y_hat)^2, dim = 2))
```

If the groups have different number of observations you could pad to the maximum number of observations and have another vector giving the counts. For example:

```
y_hat <- torch_randn(100, 10)
y <- torch_randn(100, 10)
y_c <- torch_randint(0, 100, size = 100)

loss <- torch_median(torch_sum((y - y_hat)^2, dim = 2)/y_c)
```

Note that optimizing for the median might be hard as it's not really a differentiable function.

**Comment by daniel-althoff** [comment]
Thanks for the valuable insight!

**Comment by dfalbel** [comment]
Thanks @daniel-althoff ! Happy to help
